From 12349160aa2b263e4dd52f6ae8dcd8948d1507e9 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 23 Aug 2022 18:53:32 +0300
Subject: [PATCH] [SW-94864] rom: validate start address from desc

Start address from image descriptor is used to jump to execution.
However it needs to be verified that this start address is a point
in the dma-ed image to avoid jumps to invalid address range.

This patch implements the same.

Change-Id: I58b868202e2c9ffd41cebcff6f98d0fc197a1bae
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 41d45e5..05b7839 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -633,6 +633,15 @@
 	return num_crc;
 }
 
+bool arcboot_is_valid_start_addr(struct boot_desc *desc)
+{
+	if ((desc->start_addr >= desc->loadaddr) &&
+	   (desc->start_addr < desc->loadaddr + desc->img_size))
+		return true;
+
+	return false;
+}
+
 #define PPBOOT_MAX_IMAGE_SIZE	(2 * MAX_DMA_SIZE_CRC)
 static int arcboot_rom_bringup_img_sub(int img, struct boot_desc *desc, enum fw_post_reset_cmd upd_status,
 		uint8_t cur_img)
@@ -660,6 +669,11 @@
 		return ARCBOOT_BRINGUP_ERROR;
 	}
 
+	if (!arcboot_is_valid_start_addr(desc)) {
+		arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_STARTADDR_INVALID);
+		return ARCBOOT_BRINGUP_ERROR;
+	}
+
 	security_report_img_fw(arcboot_rom_global.curr_img, desc->sec_ver);
 
 	/*
diff --git a/arch/gaudi3/arc/dev_common.h b/arch/gaudi3/arc/dev_common.h
index b8c355b..1cbf795 100644
--- a/arch/gaudi3/arc/dev_common.h
+++ b/arch/gaudi3/arc/dev_common.h
@@ -4,6 +4,7 @@
  */
 
 #include <arcboot.h>
+#include <boot_desc.h>
 
 #define GPIOS_PER_BANK			16
 #define TPM_CS_GPIO_PIN			18
@@ -32,3 +33,4 @@
 uint64_t get_pcie_pll_freq(enum pll_div_out div);
 enum dma_dest_type arcboot_rom_get_dma_dest_type(uint64_t addr);
 void arcboot_rom_halt(void);
+bool arcboot_is_valid_start_addr(struct boot_desc *desc);
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index e3b1006..12465b1 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -43,114 +43,119 @@
 	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 36,
 	PROGRESS_BRINGUP_PCI_PRI = 37,
 	PROGRESS_BRINGUP_PCI_PRI_LOADADDR_INVALID = 38,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 39,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 40,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 41,
-	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 42,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 43,
-	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 44,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 45,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 46,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 47,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 48,
-	PROGRESS_BRINGUP_PCI_SEC = 49,
-	PROGRESS_BRINGUP_PCI_SEC_LOADADDR_INVALID = 50,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 51,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 52,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 53,
-	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 54,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 55,
-	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 56,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 57,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 58,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 59,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 60,
-	PROGRESS_ARCBOOT_IN_RECOVERY = 61,
-	PROGRESS_ARCBOOT_ROM_TPM = 62,
-	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 63,
-	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 64,
-	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 65,
-	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 66,
-	PROGRESS_BRINGUP_TPM_SPI = 67,
-	PROGRESS_BRINGUP_TPM_SPI_PRI = 68,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_LOADADDR_INVALID = 69,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 70,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 71,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 72,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 73,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 74,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 75,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 76,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 77,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 78,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 79,
-	PROGRESS_BRINGUP_TPM_SPI_SEC = 80,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_LOAD_ADDR_INVALID = 81,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 82,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 83,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 84,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 85,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 86,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 87,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 88,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 89,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 90,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 91,
-	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 92,
-	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 93,
-	PROGRESS_PPBOOT_START = 94,
-	PROGRESS_PPBOOT_FINISH = 95,
-	PROGRESS_BRINGUP_RAM_PRI = 96,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 97,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 98,
-	PROGRESS_BRINGUP_RAM_SEC = 99,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 100,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 101,
-	PROGRESS_ARCBOOT_RAM_START = 102,
-	PROGRESS_ARCBOOT_RECOVERY = 103,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 104,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 105,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 106,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 107,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 108,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 109,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 110,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 111,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 112,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 113,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 114,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 115,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 116,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 117,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 118,
-	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 119,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 120,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 121,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 122,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 123,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 124,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 125,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 126,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 127,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 128,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 129,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 130,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 131,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 132,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 133,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 134,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 135,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 136,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 137,
-	PROGRESS_ARCBOOT_EROM_IMG0_LOADADDR_INVALID = 138,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 139,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 140,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 141,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 142,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 143,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 144,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 145,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 146,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 147,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 148,
+	PROGRESS_BRINGUP_PCI_PRI_STARTADDR_INVALID = 39,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 40,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 41,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 42,
+	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 43,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 44,
+	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 45,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 46,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 47,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 48,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 49,
+	PROGRESS_BRINGUP_PCI_SEC = 50,
+	PROGRESS_BRINGUP_PCI_SEC_LOADADDR_INVALID = 51,
+	PROGRESS_BRINGUP_PCI_SEC_STARTADDR_INVALID = 52,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 53,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 54,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 55,
+	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 56,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 57,
+	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 58,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 59,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 60,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 61,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 62,
+	PROGRESS_ARCBOOT_IN_RECOVERY = 63,
+	PROGRESS_ARCBOOT_ROM_TPM = 64,
+	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 65,
+	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 66,
+	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 67,
+	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 68,
+	PROGRESS_BRINGUP_TPM_SPI = 69,
+	PROGRESS_BRINGUP_TPM_SPI_PRI = 70,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_LOADADDR_INVALID = 71,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_STARTADDR_INVALID = 72,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 73,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 74,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 75,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 76,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 77,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 78,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 79,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 80,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 81,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 82,
+	PROGRESS_BRINGUP_TPM_SPI_SEC = 83,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_LOADADDR_INVALID = 84,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_STARTADDR_INVALID = 85,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 86,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 87,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 88,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 89,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 90,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 91,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 92,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 93,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 94,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 95,
+	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 96,
+	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 97,
+	PROGRESS_PPBOOT_START = 98,
+	PROGRESS_PPBOOT_FINISH = 99,
+	PROGRESS_BRINGUP_RAM_PRI = 100,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 101,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 102,
+	PROGRESS_BRINGUP_RAM_SEC = 103,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 104,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 105,
+	PROGRESS_ARCBOOT_RAM_START = 106,
+	PROGRESS_ARCBOOT_RECOVERY = 107,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 108,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 109,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 110,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 111,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 112,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 113,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 114,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 115,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 116,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 117,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 118,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 119,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 120,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 121,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 122,
+	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 123,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 124,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 125,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 126,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 127,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 128,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 129,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 130,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 131,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 132,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 133,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 134,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 135,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 136,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 137,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 138,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 139,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 140,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 141,
+	PROGRESS_ARCBOOT_EROM_IMG0_LOADADDR_INVALID = 142,
+	PROGRESS_ARCBOOT_EROM_IMG0_STARTADDR_INVALID = 143,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 144,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 145,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 146,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 147,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 148,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 149,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 150,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 151,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 152,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 153,
 };
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 8e86f49..ae296e3 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -56,6 +56,11 @@
 		return;
 	}
 
+	if (!arcboot_is_valid_start_addr(&desc)) {
+		arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_STARTADDR_INVALID);
+		return;
+	}
+
 	dst_type = arcboot_rom_get_dma_dest_type(desc.loadaddr);
 
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_DMA_START);
diff --git a/include/arcboot.h b/include/arcboot.h
index 4a29831..6141a5d 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -96,6 +96,7 @@
 enum arcboot_rom_bringup_proggress {
 	PROGRESS_BRINGUP_START,
 	PROGRESS_BRINGUP_LOADADDR_INVALID,
+	PROGRESS_BRINGUP_STARTADDR_INVALID,
 	PROGRESS_BRINGUP_TOO_BIG,
 	PROGRESS_BRINGUP_CRC_FAIL,
 	PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR,
@@ -134,6 +135,7 @@
 enum arcboot_rom_erom_img_progress {
 	PROGRESS_EROM_BRINGUP_START,
 	PROGRESS_EROM_LOADADDR_INVALID,
+	PROGRESS_EROM_STARTADDR_INVALID,
 	PROGRESS_EROM_DMA_START,
 	PROGRESS_EROM_DMA_CRC_FAILED,
 	PROGRESS_EROM_DMA_UNALIGNED_DEST_ADDR,
